src/Aspire.Hosting/Exec/ExecResourceManager.cs

```diff
@@ -0,0 +1,213 @@
+        if (targetResource is not IResourceSupportsExec targetExecResource)
+        {
+            _logger.LogWarning("Target resource '{ResourceName}' does not support exec.", _execOptions.ResourceName);
+            throw new ArgumentException($"Target resource '{_execOptions.ResourceName}' does not support exec");
+        }
+
+        var execResource = BuildResource(targetExecResource);
+
+        _logger.LogInformation("Resource '{ResourceName}' has been successfully built and added to the model resources.", execResource.Name);
+        _execResourceInitialized.SetResult(execResource);
+        return execResource;
+    }
+
+    IResource BuildResource(IResourceSupportsExec targetExecResource)
+    {
+        return targetExecResource switch
+        {
+            ProjectResource prj => BuildAgainstProjectResource(prj),
+            _ => throw new NotImplementedException(nameof(targetExecResource))
+        };
+    }
+
+    private IResource BuildAgainstProjectResource(ProjectResource project)
+    {
+        var projectMetadata = project.GetProjectMetadata();
+        var projectDir = Path.GetDirectoryName(projectMetadata.ProjectPath) ?? throw new InvalidOperationException("Project path is invalid.");
+        var (exe, args) = ParseCommand();
+
+        // unique name for the new exec resource
+        string execResourceName;
+        do
+        {
+            var shortId = Guid.NewGuid().ToString("N").Substring(0, 8);
+            execResourceName = "exec" + shortId;
+        } while (_model.Resources.Any(x => x.Name.Equals(execResourceName, StringComparisons.ResourceName)));
+
+        var executable = new ExecutableResource(execResourceName, exe, projectDir);
+        if (args is not null && args.Length > 0)
+        {
+            executable.Annotations.Add(new CommandLineArgsCallbackAnnotation((c) =>
+            {
+                c.Args.AddRange(args);
+                return Task.CompletedTask;
+            }));
+        }
+
+        // take all applicable annotations from target resource to replicate the environment
+        foreach (var annotation in project.Annotations.Where(annotation =>
+            annotation is EnvironmentAnnotation or EnvironmentCallbackAnnotation
+                       or ResourceRelationshipAnnotation or WaitAnnotation))
+        {
+            executable.Annotations.Add(annotation);
+        }
+
+        if (_execOptions.StartResource)
+        {
+            _logger.LogInformation("Exec resource '{ResourceName}' will wait until project '{Project}' starts up.", execResourceName, project.Name);
+            executable.Annotations.Add(new WaitAnnotation(project, waitType: WaitType.WaitUntilHealthy));
+        }
+
+        _logger.LogInformation("Exec resource '{ResourceName}' will run command '{Command}' with {ArgsCount} args '{Args}'.", execResourceName, exe, args?.Length ?? 0, string.Join(' ', args ?? []));
+
+        return executable;
+
+        (string exe, string[] args) ParseCommand()
```

Comment:
https://github.com/dotnet/aspire/blob/main/src/Shared/CommandLineArgsParser.cs

Reply:
added new utils API and reused